wave_kinematics/src/rotation.cpp

```diff
@@ -0,0 +1,99 @@
+#include "wave/kinematics/rotation.hpp"
+#include "wave/utils/utils.hpp"
+
+double wave::Rotation::comparision_threshold = 1e-8;
+
+wave::Rotation::Rotation() {}
+
+void wave::Rotation::setToIdentity() {
+  this->rotation_object.setIdentity();
+}
+
+void wave::Rotation::compose(const Rotation &R) {
+  this->rotation_object = this->rotation_object * R.rotation_object;
+}
+
+wave::Mat3 wave::Rotation::invert() {
+  this->rotation_object.invert();
+}
+
+wave::Vec3 wave::Rotation::rotate(const Vec3 &input_vector) {
+  return this->rotation_object.rotate(input_vector);
+}
+
+wave::Vec3 wave::Rotation::inverseRotate(const Vec3 &input_vector) {
+  return this->rotation_object.inverseRotate(input_vector);
+}
+
+wave::Vec3 wave::Rotation::rotateAndJacobian(const Vec3 &input_vector,
+                                             Mat3 &Jpoint,
+                                             Mat3 &Jparam) {
+  // Calculate the Jacobian quantities.
+  // Jacobian wrt the point is simply the rotation matrix.
+  Jpoint = this->rotation_object.matrix();
+
+  // Jacobian wrt to the lie algebra parameters is the rotated vector
+  // in skew symmetric form.
+  Vec3 point_rotated = this->rotation_object.rotate(input_vector);
+  Jparam = -1.0 * skew_symmetric_matrix(point_rotated);
+
+  return point_rotated;
+}
+
+bool wave::Rotation::fromEulerXYZ(const double rotation_x,
+                                  const double rotation_y,
+                                  const double rotation_z,
+                                  Rotation &R) {
+  double sx = sin(rotation_x);
+  double cx = cos(rotation_x);
+  double sy = sin(rotation_y);
+  double cy = cos(rotation_y);
+  double sz = sin(rotation_z);
+  double cz = cos(rotation_z);
+
+  double r00 = cz * cy;
+  double r01 = cz * sy * sx - sz * cx;
+  double r02 = cz * sy * cx + sz * sx;
+  double r10 = sz * cy;
+  double r11 = sz * sy * sx + cz * cx;
+  double r12 = sz * sy * cx - cz * sx;
+  double r20 = -sy;
+  double r21 = cy * sx;
+  double r22 = cy * cx;
+
```

Comment:
Oh, it looks like `fromEulerXYZ(a, b, c, ...)` corresponds to `kindr::EulerAnglesZyxD(c, b, a)`. See  the rotation matrix in [Kindr's unit test](https://github.com/ethz-asl/kindr/blob/master/test/rotations/EulerAnglesZyxTest.cpp#L718).

I.e. kindr's notation `EulerAnglesXyz` confusingly means something different than your `EulerXYZ`. Maybe you could add some clarifying words such as "body-fixed" or "intrinsic" or "extrinsic"  (whatever is correct) to the docs?

By the way if you change the function, the unit test can compare against the explicit "sin() + cos()" version. I think that is clearer than a pre-calculated "truth" `0.936293363584199...` of unclear origin.